### ISSUE TITLE
[NER finetune] pad_token_label_id = -100에 대한 loss 계산 질문.

### ISSUE THREAD

**Issue by DataLama** [monologg/KoELECTRA#4]
안녕하세요.
@monologg 님, KoELECTRA 모델을 공유해주셔서 정말 감사드립니다. 
koBERT 때부터 항상 발빠르게 transformers로 포팅해주셔서 정말 많은 도움이 되고 있습니다.~!

**제가 이슈를 남긴 이유는 다름이 아니라 fine-tuning task로 NER코드를 살펴보고 있는데, `pad_token_label_id = -100`에 대해서 loss를 계산하는 부분이 이해가 안되서 질문을 남깁니다.**

예를 들면, 아래와 같은 데이터가 있다고 할 때,
```
text: 금석객잔 여러분, 감사드립니다 .	
labels : ORG-B O O O
```

`processor.ner.py`를 사용해서 위 데이터를 전처리하면 아래와 같습니다.

```
tokens: tokens: [CLS] 금 ##석 ##객 ##잔 여러분 , 감사 ##드립니다 . [SEP]
input_ids: 2 465 6038 6050 6115 17067 16 12027 17282 18 3 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
attention_mask: 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
token_type_ids: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
label: -100 7 -100 -100 -100 0 -100 0 -100 0 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 -100 
```

여기서 `pad_token_label_id = -100`이 하는 역할에 대해서 다음의 두 가지로 이해했습니다.
1)  token이 tokenizer에 의해 sub-token으로 분리될 경우 sub-token의 첫번째 이외의 나머지들을 패딩.
    - e.g.) [금, ##석, ##객,  ##잔] -> [7, -100, -100, -100]
2)  [PAD] token에 대응하는 label 패딩값.

그러면 token classification의 loss를 계산할 때, 당연히 pad_token_label_id을 제외한 부분에 대해서만 loss 계산이 되어야된다고 생각했습니다.

그런데 [`ElectraForTokenClassification`의 foward](https://huggingface.co/transformers/_modules/transformers/modeling_electra.html#ElectraForTokenClassification) 부분에 대한 코드를 뜯어보니 2)번 부분은 마스킹이 되고 있지만 1)번 부분은 마스킹이 되지 않은채 계산되고 있었습니다.

```python
...

        if labels is not None:
            loss_fct = nn.CrossEntropyLoss()
            # Only keep active parts of the loss
            if attention_mask is not None:
                active_loss = attention_mask.view(-1) == 1 
                active_logits = logits.view(-1, self.config.num_labels)[active_loss]
                active_labels = labels.view(-1)[active_loss]
                loss = loss_fct(active_logits, active_labels)
            else:
                loss = loss_fct(logits.view(-1, self.config.num_labels), labels.view(-1))

```
attention_mask가 1인 부분만 active loss로 판단하면 label이 아래처럼 나올 것 같습니다.
```
label:  -100 7 -100 -100 -100 0 -100 0 -100 0 -100
```
**이렇게 나오면 CrossEntropyLoss가 정상적으로 계산되지 않을것 같은데요... 제 생각에는 단순히 attention mask를 기준으로 active_loss를 정하는 것이 아니라 `pad_token_label_id`를 기준으로 active_loss를 판단해야 될 것 같습니다.**

제가 NER-task는 많이 해보지 않아서 조심스럽게 질문드립니다ㅠㅠ  혹시 제가 놓친 부분이 있다면, 설명 부탁드릴 수 있을까요??

감사합니다~!

**Comment by monologg** [comment]
안녕하세요:)

[CrossEntropyLoss](https://pytorch.org/docs/master/generated/torch.nn.CrossEntropyLoss.html) 공식 Documentation을 보면 `ignore_index=-100`이 기본값으로 설정이 되어 있습니다. 이럴 시 label-100인 것에 대해서는 loss 계산에 반영하지 않습니다.

사실 위의 코드에서 `loss_fct=nn.CrossEntropyLoss(ignore_index=-100)`으로 명확히 명시하는게 이상적이긴 합니다.

**Comment by DataLama** [comment]
@monologg 님 
빠른 답변감사드립니다~!!!

아 그래서 label pad를 -100으로 주는거군요!!

덕분에 하나 더 배우고 갑니다~!!

@monologg 님 소소하게나마 star누르면서 응원하고 있습니다~!  앞으로도 좋은 오픈소스 기대할게요~😊

추가적인 이슈가 없으므로 본 이슈 종료하겠습니다!
